kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -274,52 +277,95 @@ private long commitWithRetry(
                 dataActions);
       }
 
-      int numTries = 0;
-      while (numTries <= maxRetries) { // leq because the first is a try, not a retry
-        logger.info("Committing transaction as version = {}.", commitAsVersion);
+      int attempt = 1;
+      while (true) {
+        // This loop exits upon either (a) commit success (return statement) or (b) commit failure,
+        // if the number of retries has been exhausted or the commit is not retryable.
+
+        logger.info(
+            "Attempting to commit transaction at version {}. Attempt {}/{}",
+            commitAsVersion,
+            attempt,
+            maxCommitAttempts);
         try {
           transactionMetrics.commitAttemptsCounter.increment();
           return doCommit(
               engine, commitAsVersion, attemptCommitInfo, dataActions, transactionMetrics);
-        } catch (FileAlreadyExistsException fnfe) {
-          logger.info(
-              "Concurrent write detected when committing as version = {}.", commitAsVersion);
-          if (numTries < maxRetries) {
-            // only try and resolve conflicts if we're going to retry
+        } catch (CommitFailedException cfe) {
+          if (!cfe.isRetryable()) {
+            // Case 1: Non-retryable exception. We must throw this.
+            logger.error(
+                "Commit attempt for version {} failed with a non-retryable exception and will not "
+                    + "be retried. Error: {}",
+                commitAsVersion,
+                cfe);
+
+            throw new RuntimeException(cfe);
```

Comment:
idea: NonRetryableCommitFailedException

Reply:
I feel like this one can just be a RuntimeException? But also if we change the commit method to not catch RuntimeExceptions -- we would never hit this in reality? which brings up the question is should we have a `isRetryable=false` case for CommitFailedException (not sure how that plays into the ccv2 design)

Reply:
Also good to address this in a separate PR